src/util/bip152.rs

```diff
@@ -0,0 +1,292 @@
+        let mut siphash = SipHasher24::new_with_keys(sip_hash_keys.0, sip_hash_keys.1);
+        siphash.write(&txid[..]);
+
+        // 3. Dropping the 2 most significant bytes from the SipHash output to make it 6 bytes.
+        let mut output = [0; 8];
+        LittleEndian::write_u64(&mut output, siphash.finish());
+        let mut id = ShortId([0; 6]);
+        id.0.copy_from_slice(&output[0..6]);
+        id
+    }
+}
+
+impl fmt::Display for ShortId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use hex;
+
+        write!(f, "{}", hex::encode(&self.0[..]))
+    }
+}
+
+impl fmt::Debug for ShortId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+impl<S: Encoder> Encodable<S> for ShortId {
+    #[inline]
+    fn consensus_encode(&self, s: &mut S) -> Result<(), Error> {
+        self.0.consensus_encode(s)
+    }
+}
+
+impl<D: Decoder> Decodable<D> for ShortId {
+    #[inline]
+    fn consensus_decode(d: &mut D) -> Result<ShortId, Error> {
+        Ok(ShortId(Decodable::consensus_decode(d)?))
+    }
+}
+
+/// A HeaderAndShortIDs structure is used to relay a block header, the short transactions
+/// IDs used for matching already-available transactions, and a select few transactions
+/// which we expect a peer may be missing.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct HeaderAndShortIds {
+    /// The header of the block being provided.
+    pub header: BlockHeader,
+    ///  A nonce for use in short transaction ID calculations.
+    pub nonce: u64,
+    ///  The short transaction IDs calculated from the transactions
+    ///  which were not provided explicitly in prefilled_txs.
+    pub short_ids: Vec<ShortId>,
+    ///  Used to provide the coinbase transaction and a select few
+    ///  which we expect a peer may be missing.
+    pub prefilled_txs: Vec<PrefilledTransaction>,
+}
+impl_consensus_encoding!(HeaderAndShortIds, header, nonce, short_ids, prefilled_txs);
+
+impl HeaderAndShortIds {
+    /// Create a new HeaderAndShortIds from a full block.
+    ///
+    /// > Nodes SHOULD NOT use the same nonce across multiple different blocks.
+    #[cfg(feature = "siphasher")]
+    pub fn from_block(block: &Block, nonce: u64, use_wtxid: bool) -> HeaderAndShortIds {
+        use util::hash::BitcoinHash;
+
+        let siphash_keys = ShortId::calculate_siphash_keys(&block.header, nonce);
+
+        HeaderAndShortIds {
+            header: block.header.clone(),
+            nonce: nonce,
+            // Provide coinbase prefilled.
+            prefilled_txs: vec![PrefilledTransaction(0, block.txdata[0].clone())],
+            short_ids: block
+                .txdata
+                .iter()
+                .skip(1) // coinbase
+                .map(|tx| {
+                    ShortId::calculate_with_siphash_keys(
+                        &match use_wtxid {
+                            true => tx.bitcoin_hash(),
```

Comment:
SendCmpct version and this should be better linked somehow.

Reply:
I think he is referring to the fact that the `use_wtxid` should be linked to the `version` field (1 -> false; 2 -> true) somehow. I'll change that to passing the version number and return an Error if it's not 1 or 2.